pkg/resources/external_volume.go

```diff
@@ -0,0 +1,717 @@
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func UpdateContextExternalVolume(ctx context.Context, d *schema.ResourceData, meta any) diag.Diagnostics {
+	client := meta.(*provider.Context).Client
+	id, err := sdk.ParseAccountObjectIdentifier(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	set := sdk.NewAlterExternalVolumeSetRequest()
+
+	if d.HasChange("comment") {
+		// not using d.GetOk as that doesn't let comments be reset to the empty string
+		set.WithComment(d.Get("comment").(string))
+	}
+
+	if d.HasChange("allow_writes") {
+		if v := d.Get("allow_writes").(string); v != BooleanDefault {
+			parsed, err := booleanStringToBool(v)
+			if err != nil {
+				return diag.FromErr(err)
+			}
+			set.WithAllowWrites(parsed)
+		} else {
+			// no way to unset allow writes - set to false as a default
+			set.WithAllowWrites(false)
+		}
+	}
+
+	if (*set != sdk.AlterExternalVolumeSetRequest{}) {
+		if err := client.ExternalVolumes.Alter(ctx, sdk.NewAlterExternalVolumeRequest(id).WithSet(*set)); err != nil {
+			return diag.FromErr(err)
+		}
+	}
+
+	if d.HasChange("storage_location") {
+		old, new := d.GetChange("storage_location")
+		oldLocations, err := extractStorageLocations(old, true)
+		if err != nil {
+			return diag.FromErr(err)
+		}
+
+		newLocations, err := extractStorageLocations(new, false)
+		if err != nil {
+			return diag.FromErr(err)
+		}
+
+		// Storage locations can only be added to the tail of the list, but can be
+		// removed at any position. Given this limitation, to keep the configuration order
+		// matching that on Snowflake the list needs to be partially recreated. For example, if a location
+		// is added in the configuration at index 5 in the list, all existing storage locations from index 5
+		// need to be removed, then the new location can be added, and then the removed locations
+		// can be added back. The storage locations lower than index 5 don't need to be modified.
+		// The removal process could be done without the above recreation, but it handles this case
+		// too so it's used for both actions.
+		longestCommonPrefix, err := LongestCommonPrefix(newLocations, oldLocations)
+		if err != nil {
+			return diag.FromErr(err)
+		}
+
+		var removedLocations []sdk.ExternalVolumeStorageLocation
+		var addedLocations []sdk.ExternalVolumeStorageLocation
+		if longestCommonPrefix == -1 {
+			removedLocations = oldLocations
+			addedLocations = newLocations
+		} else {
+			// Could +1 on the prefix here as the lists until and including this index
+			// are identical, would need to add some more checks for list length to avoid
+			// an array index out of bounds error
+			removedLocations = oldLocations[longestCommonPrefix:]
+			addedLocations = newLocations[longestCommonPrefix:]
+		}
+
+		if len(removedLocations) == len(oldLocations) {
+			// Create a temporary storage location, which is a copy of a storage location currently existing
```

Comment:
Yeh to avoid any chance of a name clash I went with this approach, as explained in https://github.com/Snowflake-Labs/terraform-provider-snowflake/pull/3106/files#r1799664807. Do you think it would be worth looking into changing that?

Reply:
Yeh to avoid any chance of a name clash I went with this approach, as explained in https://github.com/Snowflake-Labs/terraform-provider-snowflake/pull/3106/files#r1799664807. Do you think it would be worth looking into changing that?

Reply:
Got this. We could also copy the first location with a custom name and then call `updateStorageLocations`. We could remove this whole block
```			
                        if len(removedLocations) > 1 {
				// To ensure the name of the temporary storage location is unique,
				// first remove all but 1 storage locations from the list
				for _, removedStorageLocation := range removedLocations[1:] {
					err = removeStorageLocation(removedStorageLocation, client, ctx, id)
					if err != nil {
						return diag.FromErr(err)
					}
				}
			}
```

Reply:
Could do that yep, only disadvantage is we would then have the possibility of a duplicate storage location name error. For example, a volume with 2 storage locations `(location_a, temp_provider_copy_location_a)` would fail when the provider then tries to create `temp_provider_copy_location_a`. Very unlikely so maybe not an issue, happy to go with what you think 👍

Reply:
Yes, but that's the problem in each solution, right? Since we *always* have to have at least one storage location, I don't think we can avoid name clash for 100% (except recreation, but we obviously want to avoid that) - for both cases in which we use a sentinel (named for example `terraform_provider_sentinel`) or a copy of the first element with a prefix.

I just think that with a sentinel we'd have a bit simpler code, let me know what you think :)

Reply:
Hmm unless I’m missing something I think the current algorithm does avoid name clash 100% of the time. For example with a volume that has the locations `[location_a, temp_provider_copy_location_a]`, which the user changes in the config to `[location_b]`, it would:

1. Remove `temp_provider_copy_location_a` from the list, so that now the list is `[location_a]`.
2. Create the temp location based on the name of the only storage location in the list, so the list is now `[location_a, temp_provider_copy_location_a]`. No chance of name clash as there's only 1 storage location in the list.
3. Remove `location_a`, so the list is `[temp_provider_copy_location_a]`.
4. Add `location_b`, so the list is `[temp_provider_copy_location_a, location_b]`.
5. Remove `temp_provider_copy_location_a`, so the list is `[location_b]`, matching the config.

The key that makes this possible is step 1 -  first removing all elements but 1 from the list before creating the temp. Relevant part of the code is https://github.com/Snowflake-Labs/terraform-provider-snowflake/blob/afe0235e8dffcd73a1a786a55b72e83e3dd2d666/pkg/resources/external_volume.go#L383.

Reply:
Ah, I see what you mean now I think. In step 4 if `location_b`, the new storage location, matches the name of the temp location there would be a clash.

Reply:
Sounds good having a specially documented storage location name like `terraform_provider_sentinel`, will make that update, thanks for the spot!

Reply:
Yes, that's what I meant - when one of the locations to be added has the same name like the copied first element, there's a clash, sorry for not being clear. 

My biggest point is that users can even have a storage location named `terraform_provider_sentinel` but why would they ;) Either way this behavior should be documented so the users are not surprised by that.

Reply:
Makes perfect sense, cheers 🙂